drv/psc-seq-server/src/main.rs

```diff
@@ -3,48 +3,586 @@
+            // PWR_OK signals are active HIGH.
+            let ok = if ok_bits & (1 << PSU_PWR_OK_PINS[i]) != 0 {
+                Status::Good
+            } else {
+                Status::NotGood
+            };
+            match psus[i].step(now, present, ok) {
+                None => (),
+
+                Some(ActionRequired::EnableMe) => {
+                    ringbuf_entry!((now, Trace::Enabling { psu: i }));
+                    // Enable the PSU by allowing `ENABLE_L` to float low, by no
+                    // longer asserting high.
+                    sys.gpio_configure_input(
+                        PSU_ENABLE_L_PORT.pin(PSU_ENABLE_L_PINS[i]),
+                        Pull::None,
+                    );
+                }
+                Some(ActionRequired::DisableMe { attempt_snapshot }) => {
+                    if attempt_snapshot {
+                        // TODO snapshot goes here
+                    }
+                    ringbuf_entry!((
+                        now,
+                        Trace::Disabling {
+                            psu: i,
+                            present: attempt_snapshot,
+                        }
+                    ));
+
+                    // Pull `ENABLE_L` high to disable the PSU.
+                    sys.gpio_configure_output(
+                        PSU_ENABLE_L_PORT.pin(PSU_ENABLE_L_PINS[i]),
+                        OutputType::PushPull,
+                        Speed::Low,
+                        Pull::None,
+                    );
+                }
+                Some(ActionRequired::NoteStickyFault) => {
+                    ringbuf_entry!((now, Trace::StickyFault { psu: i }));
+                    // Not taking any particular action at this time.
+                }
+            }
+        }
+
+        // Wait for a pin change or timer.
+        let n = sys_recv_notification(sleep_notifications);
+        if n & notifications::TIMER_MASK != 0 {
+            // Reset our timer forward.
+            sys_set_timer(
+                Some(now.saturating_add(POLL_MS)),
+                notifications::TIMER_MASK,
+            );
+        }
+        // Ignore pin change notification bits, we just handle all the pins
+        // above. We also _enable_ the pin change interrupts at the top of the
+        // loop.
     }
 }
+
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
+enum Present {
+    #[default]
+    No,
+    Yes,
+}
+
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
+enum Status {
+    #[default]
+    NotGood,
+    Good,
+}
+
+struct Psu {
+    state: PsuState,
+}
+
+impl Psu {
+    /// Advances the PSU management state machine given the current time (`now`)
+    /// and the state of the `present` and `pwr_ok` inputs.
+    ///
+    /// This may be called at unpredictable intervals, and may be called more
+    /// than once for the same timestamp value. The implementation **must** use
+    /// `now` and the timer to control any time-sensitive operations.
```

Comment:
i kind of wonder if an ascii graphic flowchart depicting these state transitions might be useful?